Add GitLab integration

ivanayov opened the pull request:
This adds GitLab as possible SCM and enables authorisation for it as well

Signed-off-by: Ivana Yovcheva <iyovcheva@vmware.com>

Closes #27 

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Without auth:
<img width="895" alt="screen shot 2019-01-30 at 16 10 09" src="https://user-images.githubusercontent.com/4160133/51987057-490ade00-24aa-11e9-905b-9090c70c3b00.png">

<img width="1147" alt="screen shot 2019-01-30 at 16 10 17" src="https://user-images.githubusercontent.com/4160133/51987070-4f00bf00-24aa-11e9-9305-6c4af76124e4.png">

<img width="1530" alt="screen shot 2019-01-30 at 16 48 07" src="https://user-images.githubusercontent.com/4160133/51989225-4068d680-24af-11e9-9e47-b75040057ff0.png">

===================================
With auth:

<img width="1635" alt="screen shot 2019-01-31 at 16 17 05" src="https://user-images.githubusercontent.com/4160133/52059966-f138a980-2573-11e9-9691-8daa7ed04812.png">

===================================
With TLS:

<img width="1555" alt="screen shot 2019-02-07 at 18 24 01" src="https://user-images.githubusercontent.com/4160133/52426020-dcc15780-2b05-11e9-9d75-f4f025f7bf46.png">

<img width="1331" alt="screen shot 2019-02-07 at 13 57 08" src="https://user-images.githubusercontent.com/4160133/52426037-e5b22900-2b05-11e9-92d4-7184d11da9ed.png">


## Checklist:

I have:

- [x] checked my changes follow the style of the existing code / OpenFaaS repos
- [x] updated the documentation and/or roadmap in README.md
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [x] added unit tests

ivanayov commented on pkg/stack/stack.go:
```diff
@@ -50,6 +52,24 @@ func Apply(plan types.Plan) error {
+
+	}
+
+	stackErr := generateTemplate("stack", plan, authConfig{
```
I wonder whether the selection should be done based on `oauth_provider`. This simplifies the code from extra configuration and is easier for users, but may be confusing if Auth is not enabled.

alexellis replied:
Can you explain?

ivanayov replied:
It's about this line: https://github.com/openfaas-incubator/ofc-bootstrap/pull/46/files#diff-b8b32a5985a630c24f4fb2f6abf18d92R61

One may use GitLab without OAuth and then this check would be confusing. 
Do you think it's better to add this instead to the `init.yml`:
```
scm: github
#scm: gitlab
```
?

alexellis commented on README.md:
```diff
@@ -158,13 +170,18 @@ When ofc-bootstrap has completed and you know the IP of your LoadBalancer:
+#### Configure the GitHub / GitLab App webhook
 
-Now over on GitHub enter the URL for webhooks:
+Now over on GitHub / GitLab enter the URL for webhooks:
```
@martindekov do we explain in either here or the main repo how to obtain all the config and settings for GitLab?

martindekov replied:
Sorry Alex I saw that comment just now .. otherwise yes in [here](https://github.com/openfaas/openfaas-cloud/blob/master/docs/GITLAB.md)

alexellis replied:
@ivanayov please add Martin's link above to this PR asap. 
https://github.com/openfaas/openfaas-cloud/blob/master/docs/GITLAB.md

ivanayov replied:
@alexellis It's added now below

alexellis commented on init.yaml:
```diff
@@ -47,11 +47,22 @@ secrets:
       - name: "private-key"
         value_from: "~/Downloads/ofc-bootstrap-test.2018-12-23.private-key.pem"
     namespace: "openfaas-fn"
+  # Enter your GitLab Wekhook secret and API token
```
`Wekhook` webhook

ivanayov commented:
So we tested it both with and without Auth now

alexellis commented:
@ivanayov can you confirm?

- tested with auth on/off
- tls on/off

ivanayov commented:
@alexellis I confirm with auth on/off.
Now testing with TLS

ivanayov commented:
Build with TLS on and issuer prod:

<img width="1331" alt="screen shot 2019-02-07 at 13 57 08" src="https://user-images.githubusercontent.com/4160133/52410402-e97f8480-2ae0-11e9-9035-18f84c3354c5.png">

alexellis commented on pkg/types/types.go:
```diff
@@ -6,19 +6,22 @@ import (
+	TLSConfig            TLSConfig                `yaml:"tls_config"`
+	Slack                Slack                    `yaml:"slack"`
+	Ingress              string                   `yaml:"ingress"`
+	OAuthProvider        string                   `yaml:"oauth_provider"`
```
I feel this should all be a sub-element of the OAuth property, do you agree?

ivanayov replied:
Yes, it's more clear

alexellis commented on pkg/types/types.go:
```diff
@@ -78,3 +81,8 @@ type TLSConfig struct {
+
+type Gitlab struct {
+	GitLabInstance  string `yaml:"gitlab_instance"`
+	InstallationTag string `yaml:"installation_tag"`
```
This should not be configurable, it's `openfaas-cloud` as a default.

ivanayov replied:
Ok, moving it to the default value. In the GitLab README this was explained as configurable allowing to use custom tag value. But makes sense to stick to the default option.

alexellis commented on templates/stack.yml:
```diff
@@ -0,0 +1,204 @@
+    environment_file:
+      - gateway_config.yml
+      - github.yml
+    secrets:
```
Just curious, what ideas do you have on how to avoid templating this file?

alexellis replied:
Is it just for the single secret that we need this?

ivanayov replied:
Yes, it's only for the secret.
I don't think OpenFaaS allows a function to use a secret without having it declared. 
One optional solution is to use secrets filters in openfaas-cloud as we're going to do here.
Another option is to always create the secret, but it's not good solution I think.

`stack.yml` will also be needed for the Swarm implementation.

zeerorg replied:
If this is an issue with the git-tar function we can only deploy the function itself i.e., after checking the oauth provider deploy just the git-tar function, by create a stack.yml file with just the git-tar function with gitlab-api-token. so there won't be a need to generate stack.yml

zeerorg replied:
Or add this function to gitlab.yml file

ivanayov replied:
Thanks for the proposal @zeerorg ! This is a good option. Would cause more refactoring once templating is moved to openfaas-cloud but for now this would work I think.

zeerorg commented on example.init.yaml:
```diff
@@ -101,14 +114,29 @@ root_domain: "myfaas.club"
+  gitlab_instance: "https://gitlab.o6s.io/"
+  ## The OpenFaaS Cloud instance will build repositories with the given tag.
+  #  Can be custom populated with required value.
+  installation_tag: "openfaas-cloud"
```
I don't think this is actually doing anything, should be referenced in plan type and be generated with template

ivanayov replied:
Hm, I think I removed it. Seems like it's forgotten here

zeerorg replied:
👌

ivanayov commented:
@alexellis I addressed all comments and tested with TLS and OAuth.
Removed the `stack.yml` from templates. @zeerorg  gave very good proposal, so the gitlab-specific secret is added to `git-tar` with the gitlab deployment now.

I specified `scm` config in the init.yaml and removed `oauth_provider` as it might be confusing when auth is disabled. In code `oauth_provider` is now taken from the `scm`.

![screen shot 2019-02-22 at 18 59 15](https://user-images.githubusercontent.com/4160133/53258011-3adc6600-36d4-11e9-819d-c5d97eb10b98.png)

With the dashboard after authorising I have `Internal server error creating JWT`. Mike Whitehead wrote in slack that he had the same issue and I think it's not ofc-bootstrap related. Might be some problem with the OAuth app. Both `jwt-private-key` and `jwt-public-key` are properly created.

alexellis commented:
Sounds like we have what we need now 👍 thank you

The last part will be to look at the changes that @martindekov made in #72 and replicate that hear too. A rebase is now needed too. 

Alex

alexellis commented:
Thank you for rebasing.

martindekov commented:
Hey @ivanayov in [types.go](https://github.com/openfaas-incubator/ofc-bootstrap/blob/4191343ede7b8e41de0f42d0c02fde187e2c5376/pkg/types/types.go#L14) I have added the filter `scm_gitlab` and in [main.go](https://github.com/openfaas-incubator/ofc-bootstrap/blob/4191343ede7b8e41de0f42d0c02fde187e2c5376/main.go#L641) you need to derive the feature (`scm_gitlab` or `scm_github`) depending on the configuration you chose in `init.yaml`. You can check the tests for more context I think you might need to fix them in case they fail.

Otherwise if it is too much I can pick that up for you no problems 👍

alexellis commented:
I'll merge Martin please can you take over the final piece?

Alex
